CIPs/0017.md

```diff
@@ -0,0 +1,181 @@
+```
+GetProofInventory - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
+    `highestEpoch` is the current highestEpoch the client knows of.
+    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
+                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
+    `requestedProofs`: Set of proofs' metadata requested by the client
+  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
+  Both responses are subject to message size limits limiting the number of proofs/headers.
+EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
+```
+
+The client will call `GetProofInventory`, either during the sync or pipelined within the handshake. This information will then be stored for each peer, and before each request the client will assess the "best peer", where "best" here would be defined as the minimum number of proofs & headers needed to sync from the client's epoch to `currEpoch`. This calculation could work as follows: 
+
+`numEpochs`: Number of epochs we need to sync, the difference between the current epoch of the canonical chain (`currEpoch`) and the client's highest epoch `clientEpoch` = `currEpoch - clientEpoch`
+
+`optimalPath`: For each peer, compute the optimal path of proofs available based on the client's available VKs and known val sets.
+
+`proofRange` : For each proof, the difference between `LastEpoch` and `FirstEpoch`'s indices.
+
+`peerScore`: This is the total number of proofs and headers needed to sync by this peer. For each peer, subtract the range of each proof in `optimalPath` from `numEpochs`, the result is the number of necessary remaining headers. Add `optimalPath`.length to this for the peer's score. 
+
+Whichever peer has the lowest score indicates the shortest path (best set of proofs) for our client. This could also be short-circuited if the optimal score is provided. Once the best peer is chosen, the client requests the set of proofs and headers from this peer that it has assessed as optimal.
+
+Option 2:
+```
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, knownValSets *big.Int, maxResponses uint)
+    `knownValSets` is a bitmap of the last ~240 or so epochs that the client knows the val set for
+    `maxResponses` indicates the maximum number of (proofs + headers) the client is willing to accept
```

Comment:
We should generalize this concept to a "verification cost" metric based on the estimates that @psivesely is discussing, plus a cost metric in terms of data usage, which we could include in a unified number of separate it out.

I am not sure whether this max cost function will work as intended in the case where proofs are well distributed. My thinking is that the client will always want to sync (i.e. it will never decide to stay offline because the cost to sync is too high) so if a server cannot fullfil it's request becasuse the provided max cost is too restrictive, we can either assume that another peer will have a lower cost, and "shop around", or if that is a bad assumtion, the client will simply raise the max cost, having wasted a round trip. As a result, I'm not sure we should include this parameter.

Reply:
Especially if we're going with option 1 (proof inventory) this added parameter is unnecessary.

If we do add it, though, this idea of a verification cost makes sense to me and is an improvement to this idea.